guarddog/analyzer/sourcecode/__init__.py

```diff
@@ -1,33 +1,106 @@
 import os
 import pathlib
+from dataclasses import dataclass
+from typing import Optional, Iterable
 
 import yaml
 from yaml.loader import SafeLoader
 
 from guarddog.ecosystems import ECOSYSTEM
 
 current_dir = pathlib.Path(__file__).parent.resolve()
-rule_file_names = list(
-    filter(
-        lambda x: x.endswith('yml'),
-        os.listdir(current_dir)
-    )
-)
 
-SOURCECODE_RULES = {
-    ECOSYSTEM.PYPI: list(),
-    ECOSYSTEM.NPM: list(),
-}  # type: dict[ECOSYSTEM, list[dict]]
 
-for file_name in rule_file_names:
+# These data class aim to reduce the spreading of the logic
+# Instead of using the a dict as a structure and parse it difffently depending on the type
+@dataclass
+class SourceCodeRule:
+    """
+    Base class for source code rules
+    """
+    id: str
+    file: str
+
+
+@dataclass
+class YaraRule(SourceCodeRule):
+    """
+    Yara rule just reimplements base
+    """
+    pass
+
+
+@dataclass
+class SempgrepRule(SourceCodeRule):
+    """
+    Semgrep rule are language specific
+    Content of rule in yaml format is accessible through rule_content
+    """
+    description: str
+    ecosystem: ECOSYSTEM
+    rule_content: dict
+
+
+def get_sourcecode_rules(
+    ecosystem: ECOSYSTEM, kind: Optional[type] = None
+) -> Iterable[SourceCodeRule]:
+    """
+    This function returns the source code rules for a given ecosystem and kind.
+    Args:
+        ecosystem: The ecosystem to filter for if rules are ecosystem specific
+        kind: The kind of rule to filter for
+    """
+    for rule in SOURCECODE_RULES:
+        if not (getattr(rule, "ecosystem", ecosystem) == ecosystem):
+            continue
+        if kind and not isinstance(rule, kind):
+            continue
+        yield rule
+
+
+SOURCECODE_RULES: list[SourceCodeRule] = list()
+
+semgrep_rule_file_names = list(
+    filter(lambda x: x.endswith("yml"), os.listdir(current_dir))
+)
+# all yml files placed in the sourcecode directory are loaded as semgrep rules
+# refer to README.md for more information
+for file_name in semgrep_rule_file_names:
     with open(os.path.join(current_dir, file_name), "r") as fd:
         data = yaml.load(fd, Loader=SafeLoader)
         for rule in data["rules"]:
             for lang in rule["languages"]:
+                ecosystem = None
                 match lang:
                     case "python":
-                        if rule not in SOURCECODE_RULES[ECOSYSTEM.PYPI]:
-                            SOURCECODE_RULES[ECOSYSTEM.PYPI].append(rule)
+                        ecosystem = ECOSYSTEM.PYPI
                     case "javascript" | "typescript" | "json":
-                        if rule not in SOURCECODE_RULES[ECOSYSTEM.NPM]:
-                            SOURCECODE_RULES[ECOSYSTEM.NPM].append(rule)
+                        ecosystem = ECOSYSTEM.NPM
+                    case _:
+                        continue
+
+                # avoids duplicates when multiple languages are supported by a rule
```

Comment:
I think this is really for the `"javascript" | "typescript" | "json"` case, right?

Reply:
actually, the semgrep rule can have multiple languages, in here we're just doing the parsing and adding it to our memory ruleset.

for instance, `shady-links` supports: javascript, typescript, python, and soon go.
so, we need this thing to add the rule once per supported language (1 for javascript, 1 go python, 1 for go)